File: 237G_Computational/Projects/Project4/Test/Project4/Project4/Question1.cpp
#include "Question1.h"
#include <algorithm>
#include <math.h>

using namespace std;

void Question1()
{
	double sigma = 0.24;
	double r = 0.05;
	double s0 = 32;
	double K = 30;
	double time = 0.5;

	vector<int> noOfSteps{ 10,20,40,80,100,200,500 };
	vector<vector<double>> answers;
	
	//Part a
	vector<double> answersa, answersb, answersc, answersd;
	for (int count = 0; count < noOfSteps.size(); count++)
	{
		double delta = time / noOfSteps[count];
		double ansa = Parta(r, delta, sigma, noOfSteps[count], s0, K);
		answersa.push_back(ansa);

		double ansb = Partb(r, delta, sigma, noOfSteps[count], s0, K);
		answersb.push_back(ansb);

		double ansc = Partc(r, delta, sigma, noOfSteps[count], s0, K);
		answersc.push_back(ansc);

		double ansd = Partd(r, delta, sigma, noOfSteps[count], s0, K);
		answersd.push_back(ansd);
	}
	answers.push_back(answersa);
	answers.push_back(answersb);
	answers.push_back(answersc);
	answers.push_back(answersd);

	//To plot graphs
	writeToExcel("Q1.csv",answers);
	//
	
	cout << endl << "All the prices for the 4 types and various values of N are"<<endl;
	writeToConsole(answers);

	//Convergence value
	double bs_callPrice = BlackScholes_Call(s0, time, K, r, sigma);
	cout << endl << "Convergence price ,i.e., Black Scholes price = " << bs_callPrice << endl;
}

double Parta(double r, double delta, double sigma, int noOfSteps, double s0, double K)
{
	double c = (exp(-1.0 * r * delta) + exp((r + pow(sigma, 2))*delta)) / 2.0;
	double d = c - sqrt(pow(c, 2) - 1);
	double u = 1.0 / d;
	double prob = (exp(r*delta) - d) / (u - d);
	return PriceOption(r, noOfSteps, delta, prob, u, d, s0, K);
}

double Partb(double r, double delta, double sigma, int noOfSteps, double s0, double K)
{
	double d = exp(r*delta)*(1 - sqrt(exp(pow(sigma, 2)*delta) - 1));
	double u = exp(r*delta)*(1 + sqrt(exp(pow(sigma, 2)*delta) - 1));
	double prob = 0.5;
	return PriceOption(r, noOfSteps, delta, prob, u, d, s0, K);
}

double Partc(double r, double delta, double sigma, int noOfSteps, double s0, double K, char optionType)
{
	double d = exp((r - (pow(sigma, 2) / 2))*delta - sigma*sqrt(delta));
	double u = exp((r - (pow(sigma, 2) / 2))*delta + sigma*sqrt(delta));
	double prob = 0.5;
	return PriceOption(r, noOfSteps, delta, prob, u, d, s0, K, optionType);
}

double Partd(double r, double delta, double sigma, int noOfSteps, double s0, double K)
{
	double d = exp(-1.0 * sigma*sqrt(delta));
	double u = exp(sigma*sqrt(delta));
	double prob = 0.5 + 0.5 * ((r - (pow(sigma, 2) / 2))*sqrt(delta)) / sigma;
	return PriceOption(r, noOfSteps, delta, prob, u, d, s0, K);
}

double PriceOption(double r, int noOfSteps, double delta, double prob, double u, double d, double s0, double K, char optionType)
{
	double price = 0.0;
	for (int count = 0; count <= noOfSteps; count++)
	{
		double st = s0 * pow(u, count) * pow(d, noOfSteps - count);
		double priceVal = optionType == 'c' ? max(st - K, 0.0) : max(K - st, 0.0);
		double bnlProb = BnlProb(noOfSteps, count, prob);
		price += bnlProb*priceVal;
	}
	return (exp(-1.0 * r * noOfSteps * delta) * price);
}

double BnlProb(unsigned n, unsigned k, double prob)
{
	return exp(lgamma(n + 1) - lgamma(k + 1) - lgamma(n - k + 1))*pow(prob, k)*pow(1 - prob, n - k);
}

double BlackScholes_Call(double S0, double T, double X, double r, double sig)
{
	double d1 = (log(S0 / X) + (r + pow(sig, 2) / 2)*T) / (sig*sqrt(T));
	double d2 = d1 - sig*sqrt(T);

	double test = exp(-1.0 * r * T);
	double price = S0 * NCap(d1) - X * test * NCap(d2);

	return(price);
}


double NCap(double x)
{
	if (x < 0)
	{
		return 1 - NCap(-1.0 * x);
	}

	// constants
	double d1 = 0.0498673470;
	double d2 = 0.0211410061;
	double d3 = 0.003277623;
	double d4 = 0.0000380036;
	double d5 = 0.0000488906;
	double d6 = 0.0000053830;

	// Save the sign of x
	double val = 1 + d1*x + d2*pow(x, 2) + d3*pow(x, 3) + d4*pow(x, 4)
		+ d5*pow(x, 5) + d6*pow(x, 6);
	return 1.0 - 0.5 * pow(val, -1.0 * 16);
}

File: 237G_Computational/Projects/Project4/Python/Question1.py
import math
import numpy as np
import pandas as pd


def question1():
    sigma = 0.24;
    r = 0.05;
    s0 = 32;
    K = 30;
    time = 0.5;

    no_of_steps = [10, 20, 40, 80, 100, 200, 500]
    answers = np.zeros((4,len(no_of_steps)))


    for count in range(0,len(no_of_steps)):
        delta = time / no_of_steps[count];
        answers[0,count] = Parta(r, delta, sigma, no_of_steps[count], s0, K)
        answers[1,count] = Partb(r, delta, sigma, no_of_steps[count], s0, K)
        answers[2,count] = Partc(r, delta, sigma, no_of_steps[count], s0, K, "c")
        answers[3,count] = Partd(r, delta, sigma, no_of_steps[count], s0, K)

    ans_pd = pd.DataFrame(answers,index=["Method1","Method2","Method3","Method4"],columns=no_of_steps)
    print(ans_pd)

def Parta(r, delta, sigma, no_of_steps, s0, K):
    c = (math.exp(-1.0 * r * delta) + math.exp((r + pow(sigma, 2))*delta)) / 2.0;
    d = c - math.sqrt(pow(c, 2) - 1);
    u = 1.0 / d;
    prob = (math.exp(r*delta) - d) / (u - d);
    return PriceOption(r, no_of_steps, delta, prob, u, d, s0, K)


def Partb(r, delta, sigma, no_of_steps, s0, K):
    d = math.exp(r*delta)*(1 - math.sqrt(math.exp(pow(sigma, 2)*delta) - 1));
    u = math.exp(r*delta)*(1 + math.sqrt(math.exp(pow(sigma, 2)*delta) - 1));
    prob = 0.5;
    return PriceOption(r, no_of_steps, delta, prob, u, d, s0, K);


def Partc(r, delta, sigma, noOfSteps, s0, K, optionType="c"):
    d = math.exp((r - (pow(sigma, 2) / 2))*delta - sigma*math.sqrt(delta));
    u = math.exp((r - (pow(sigma, 2) / 2))*delta + sigma*math.sqrt(delta));
    prob = 0.5;
    return PriceOption(r, noOfSteps, delta, prob, u, d, s0, K, optionType);


def Partd(r, delta, sigma, no_of_steps, s0, K):
    d = math.exp(-1.0 * sigma*math.sqrt(delta))
    u = math.exp(sigma*math.sqrt(delta))
    prob = 0.5 + 0.5 * ((r - (pow(sigma, 2) / 2))*math.sqrt(delta)) / sigma
    return PriceOption(r, no_of_steps, delta, prob, u, d, s0, K)


def PriceOption(r, no_of_steps, delta, prob, u, d, s0, K, option_type = "c"):
    time = np.array(range(0,(no_of_steps + 1)))

    st = s0 * np.power(u,time) * np.power(d,(no_of_steps-time))
    price = np.amax(np.column_stack((st-K,np.array([0]*(no_of_steps+1)))),axis=1) if option_type == "c" else\
            np.amax(np.column_stack((K-st,np.array([0]*(no_of_steps+1)))),axis=1)

    bnl_probs = [bnl_prob(no_of_steps, t, prob) for t in time]

    return sum(price * bnl_probs) * math.exp(- r * no_of_steps * delta)


def bnl_prob(n, k, prob):
    return nCr(n,k) * (prob**k) * ((1-prob)**(n-k))
    #return math.exp(sc.gamma(n + 1) - sc.gamma(k + 1) - sc.gamma(n - k + 1))* pow(prob, k)*pow(1 - prob, n - k)


def nCr(n,r):
    f = math.factorial
    return f(n) // f(r) // f(n-r)